Cross-crate RPITIT loses bounds

### Cross-crate

I tried this code:

`a/src/lib.rs`
```rust
#![feature(return_position_impl_trait_in_trait)]

pub trait A {
    fn get_iter(&self) -> impl Iterator<Item = ()>;
}
```
`b/src/lib.rs`
```rust
use a::A;

pub fn b(a: &impl A) -> usize {
    a.get_iter().count()
}
```

Compiled fine with `1.72.0-nightly (5ea666864 2023-06-27)`, but gives the following error with `1.72.0-nightly (5bd28f5ea 2023-06-28)`:

```
error[E0599]: `impl ?Sized` is not an iterator
 --> src/lib.rs:6:18
  |
4 |     a.get_iter().count()
  |                  ^^^^^ `impl ?Sized` is not an iterator
  |
  = note: the following trait bounds were not satisfied:
          `impl ?Sized: Iterator`
          which is required by `&mut impl ?Sized: Iterator`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `b` (lib) due to previous error
```

### Single crate

My initial attempt at minimizing the case shown that the single-crate equivalent:

`c/src/lib.rs`
```rust
#![feature(return_position_impl_trait_in_trait)]

pub trait A {
    fn get_iter(&self) -> impl Iterator<Item = ()>;
}

pub fn b(a: &impl A) -> usize {
    a.get_iter().count()
}
```

compiles just fine.

### Meta

`rustc --version --verbose`:
```
rustc 1.72.0-nightly (5bd28f5ea 2023-06-28)
binary: rustc
commit-hash: 5bd28f5eac1ba3569bfa8d49ec3f5acbdfdff7a0
commit-date: 2023-06-28
host: x86_64-unknown-linux-gnu
release: 1.72.0-nightly
LLVM version: 16.0.5
```

EDIT: Reduced reproduction code.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"compiler-errors"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_metadata/src/rmeta/encoder.rs
+++ b/compiler/rustc_metadata/src/rmeta/encoder.rs
@@ -1447,6 +1447,9 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
                     .is_type_alias_impl_trait
                     .set(def_id.index, self.tcx.is_type_alias_impl_trait(def_id));
             }
+            if let DefKind::ImplTraitPlaceholder = def_kind {
+                self.encode_explicit_item_bounds(def_id);
+            }
             if tcx.impl_method_has_trait_impl_trait_tys(def_id)
                 && let Ok(table) = self.tcx.collect_return_position_impl_trait_in_trait_tys(def_id)
             {

--- a/tests/ui/impl-trait/in-trait/foreign.rs
+++ b/tests/ui/impl-trait/in-trait/foreign.rs
@@ -14,6 +14,10 @@ impl Foo for Local {
     fn bar(self) -> Arc<String> { Arc::new(String::new()) }
 }
 
+fn generic(f: impl Foo) {
+    let x = &*f.bar();
+}
+
 fn main() {
     // Witness an RPITIT from another crate.
     let &() = Foreign.bar();

